### ISSUE TITLE
Unexpected unification condition when rewriting

### ISSUE THREAD

**Issue by QGarchery** [Deducteam/lambdapi#713]
When rewriting a subterm such that the type of the whole term depends on it (dependent type), it can happen that a new unification goal is created. Is it to be expected ? If so, when is it added and why ? See [this example](https://github.com/Deducteam/lambdapi/files/6747407/rewrite_dependent.txt) for a minimal encoding showing the added unification goal.

**Comment by fblanqui** [comment]
Hi @QGarchery . Note that your lemma can be proved by using reflexivity only, but I guess you want to show a problem with rewrite. The problem is that, to rewrite (succ x) into (S x), we need to abstract over (succ x). The goal then get refined by (@=_ind nat (succ x) (S x) equality (λ z: ε nat, @= (Q (S x)) (ind_ℕ Q qO qS z) (ind_ℕ Q qO qS (S x)))). But (ind_ℕ Q qO qS z) is of type (Q z) and not of type (Q (S x)). You get the same problem in Coq:
```
Definition z := 0.
Lemma applyInd Q q0 qS : nat_ind Q q0 qS z = nat_ind Q q0 qS 0.
  assert (z = 0).
  reflexivity.
  rewrite H.

Error: Abstracting over the term "z" leads to a term
fun n : nat => nat_ind Q q0 qS n = nat_ind Q q0 qS 0
which is ill-typed.
Reason is: Illegal application: 
The term "@eq" of type "forall A : Type, A -> A -> Prop"
cannot be applied to the terms
 "Q n" : "Prop"
 "nat_ind Q q0 qS n" : "Q n"
 "nat_ind Q q0 qS 0" : "Q 0"
The 3rd term has type "Q 0" which should be coercible to "Q n".
```
The problem is that LP doesn't fail. It doesn't see that the constraint z ≡ S x is not satisfiable. We should try to fix this.

**Comment by fblanqui** [comment]
With the proposed PR, your file now fails with:
```
[z]
and
[S x]
are not unifiable.
Unification goals are unsatisfiable.
```

**Comment by fblanqui** [comment]
Does it solve this issue?

**Comment by QGarchery** [comment]
Hi @fblanqui, thank you for the reply !

Yes that answers my questions and I think the PR solves the issue.

I had noticed that we can prove the goal by simply doing `reflexivity;` and I put that solution in the file (as a side note, my goals are duplicated after the `have` tactic). What I did not realize is that we can have a proof that follows the same steps as my original proof. That proof uses the convertibility of terms instead of their equality and consists in doing  `have e2 : ε (ind_ℕ Q qO qS O = ind_ℕ Q qO qS O);`.
